### PR TITLE
Fix/Real Trade Alpaca

### PR DESCRIPTION
**Pull request by prithviraj-gotepatil-algobulls**
- different variables for pnl table
- timeout error handling for delete request
- new parameter for initial funds added to BT and PT
- added real trade
- removed code duplication for PT RT and BT
- added a broker id parameter for PT RT and BT
- added a broker details parameter for PT RT and BT
- allowed stop for RealTrade
- new format for start and end timestamp input (with and without timezone)

### CODE REVIEW

File: pyalgotrading/algobulls/api.py [hunk 1]
```diff
@@ -256,10 +256,13 @@ def delete_previous_trades(self, strategy: str):
             response: response from api
 
         Info: ENDPOINT
-           `DELETE` v3/build/python/user/strategy/deleteAll?strategyId={strategy}
+           `DELETE` https://api.algobulls.com/v3/build/python/user/strategy/deleteAll?strategyId=<strategy_code>
         """
+        base_url = 'https://api.algobulls.com/'
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/api.py:261]: why is this redefined

File: pyalgotrading/algobulls/api.py [hunk 2]
```diff
@@ -256,9 +262,12 @@ def delete_previous_trades(self, strategy: str):
...
         endpoint = f'v3/build/python/user/strategy/deleteAll?strategyId={strategy}'
+        url = f'{base_url}{endpoint}'
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/api.py:269]: do you really need this?

File: pyalgotrading/algobulls/connection.py [hunk 3]
```diff
@@ -183,10 +184,24 @@ def delete_previous_trades(self, strategy):
         Args:
             strategy: Strategy code
         """
+
         response = self.api.delete_previous_trades(strategy)
-        if response['data'] != 'success':
-            print(f'Deletion of previous trade data for strategy "{strategy}" may not have been successful. Details: {response}')
+        print(response)
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:189]: cleanup ?

File: pyalgotrading/algobulls/api.py [hunk 4]
```diff
@@ -256,10 +256,13 @@ def delete_previous_trades(self, strategy: str):
             response: response from api
 
         Info: ENDPOINT
-           `DELETE` v3/build/python/user/strategy/deleteAll?strategyId={strategy}
+           `DELETE` https://api.algobulls.com/v3/build/python/user/strategy/deleteAll?strategyId=<strategy_code>
```
[the remark this answers is not in the document]
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/api.py]: done

File: pyalgotrading/algobulls/api.py (hunk 2 above)
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/api.py]: done

File: pyalgotrading/algobulls/connection.py (hunk 3 above)
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/connection.py]: done

File: pyalgotrading/algobulls/api.py [hunk 5]
```diff
@@ -256,9 +262,10 @@ def delete_previous_trades(self, strategy: str):
             response: response from api
 
         Info: ENDPOINT
-           `DELETE` v3/build/python/user/strategy/deleteAll?strategyId={strategy}
+           `DELETE` v3/build/python/user/strategy/deleteAll?strategyId=<strategy_code>
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/api.py:265]: why this change?

File: pyalgotrading/algobulls/connection.py [hunk 6]
```diff
@@ -271,7 +279,94 @@ def get_report(self, strategy_code, trading_type, report_type, render_as_datafra
             print('Report not available yet. Please retry in sometime')
             # return response
 
-    def backtest(self, strategy=None, start=None, end=None, instruments=None, lots=1, parameters=None, candle=None, mode=StrategyMode.INTRADAY, delete_previous_trades=True, **kwargs):
+    def get_pnl_report_table(self, strategy_code, trading_type):
+        """
+            Fetch Back Testing Profit & Loss details
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:284]: "Back Testing" not right as this function is used by BT/PT/RT all

File: pyalgotrading/algobulls/connection.py [hunk 7]
```diff
@@ -709,11 +764,116 @@ def get_papertrading_report_order_history(self, strategy_code):
...
+            'instrument' behaves same 'instruments'
+            'strategy_parameters' behaves same 'parameters'
+            'candle_interval' behaves same 'candle'
+            'strategy__mode' behaves same 'mode'
+
+        Returns:
+            realtrade job submission status
         """
+
+        # check if values received by new parameter names, else extract from old parameter names
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:800]: i see code duplication in realtrade; it should be similar to papertrade one; pls implement DRY

File: pyalgotrading/algobulls/api.py [hunk 8]
```diff
@@ -285,7 +291,7 @@ def set_strategy_config(self, strategy_code: str, strategy_config: dict, trading
         print('Success.')
         return key, response
 
-    def start_strategy_algotrading(self, strategy_code: str, start_timestamp: dt, end_timestamp: dt, trading_type: TradingType, lots: int) -> dict:
+    def start_strategy_algotrading(self, strategy_code: str, start_timestamp: dt, end_timestamp: dt, trading_type: TradingType, lots: int, initial_funds_virtual=1e9, broker_id: int = None, broker_details: dict = None) -> dict:
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/api.py:294]: `broker_id` parameter not required as everything is coming through `broking_details`. Please clean everywhere.

File: pyalgotrading/algobulls/api.py [hunk 9]
```diff
@@ -315,18 +317,26 @@ def start_strategy_algotrading(self, strategy_code: str, start_timestamp: dt, en
...
+                'brokingDetails': broker_details
             }
+
+            params = None
+            if trading_type in [TradingType.PAPERTRADING, TradingType.BACKTESTING]:
+                endpoint = 'v4/portfolio/strategies?isPythonBuild=true&isLive=false'
+                execute_config['initialFundsVirtual'] = initial_funds_virtual
+            elif trading_type is TradingType.REALTRADING:
+                endpoint = 'v4/portfolio/strategies?isPythonBuild=true&isLive=true'
+                execute_config['brokerId'] = broker_id
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/api.py:332]: `broker_id` parameter not required as everything is coming through `broking_details`. Please clean everywhere.

File: pyalgotrading/algobulls/api.py [hunk 10]
```diff
@@ -315,18 +317,26 @@ def start_strategy_algotrading(self, strategy_code: str, start_timestamp: dt, en
             }
             _timestamp_format = "%d-%m-%YT%H:%MZ"
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/api.py:318]: is this required anywhere?

File: pyalgotrading/algobulls/connection.py [hunk 11]
```diff
@@ -183,9 +185,15 @@ def delete_previous_trades(self, strategy):
...
-        if response['data'] != 'success':
-            print(f'Deletion of previous trade data for strategy "{strategy}" may not have been successful. Details: {response}')
+        response = {}
+        for _ in range(30):
+            response = self.api.delete_previous_trades(strategy)
+            if response.get('data') == 'success':
+                print(response.get('message'))
+                break
+            else:
+                print(f'deleting previous trades ... attempt [{_}])\n{response}\n')
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:195]: Capitalize the first word; There should be no space before `...`

File: pyalgotrading/algobulls/connection.py [hunk 12]
```diff
@@ -271,20 +279,112 @@ def get_report(self, strategy_code, trading_type, report_type, render_as_datafra
...
+            Returns:
+                Report details
+        """
+        assert isinstance(strategy_code, str), f'Argument "strategy_code" should be a string'
+
+        # Fetch the data
+        data = self.get_report(strategy_code=strategy_code, trading_type=trading_type, report_type=TradingReportType.PNL_TABLE)
+
+        # Post-processing: Cleanup & converting data to dataframe
+        column_rename_map = OrderedDict([
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:299]: move this line inside the `if` cond on the next line.

File: pyalgotrading/algobulls/connection.py [hunk 13]
```diff
@@ -271,20 +279,112 @@ def get_report(self, strategy_code, trading_type, report_type, render_as_datafra
...
+        # save as html file
+        if html_dump:
+            all_strategies = self.get_all_strategies()
+            strategy_name = all_strategies.loc[all_strategies['strategyCode'] == strategy_code]['strategyName'].iloc[0]
+            qs.reports.html(total_funds_series, title=strategy_name, output='', download_filename=f'report_{strategy_name}_{time.time():.0f}.html')
+
+        return order_report
+
+    def start_job(self, strategy=None, start=None, end=None, instruments=None, lots=1, parameters=None, candle=None, mode=StrategyMode.INTRADAY, initial_funds_virtual=1e9, delete_previous_trades=True, trading_type=None, broker_id=None,
+                  broking_details=None, **kwargs):
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:370]: `broker_id` parameter not required as everything is coming through `broking_details`. Please clean everywhere.

File: pyalgotrading/algobulls/connection.py [hunk 14]
```diff
@@ -271,20 +279,112 @@ def get_report(self, strategy_code, trading_type, report_type, render_as_datafra
...
+            end: End date-time/time
             instruments: Instrument key
             lots: Number of lots of the passed instrument to trade on
             parameters: Parameters
             candle: Candle interval
             mode: Intraday or delivery
             delete_previous_trades: Delete data for previous trades
+            initial_funds_virtual: virtual funds allotted before the backtesting starts
+            trading_type: type of trading : PT/BT/RT
+            broker_id: ID of the broker
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:386]: `broker_id` parameter not required as everything is coming through `broking_details`. Please clean everywhere.

File: pyalgotrading/algobulls/connection.py [hunk 15]
```diff
@@ -271,20 +279,112 @@ def get_report(self, strategy_code, trading_type, report_type, render_as_datafra
...
+            broking_details: details of client's broker
 
         Legacy args (will be deprecated in future release):
             'strategy_code' behaves same 'strategy'
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:390]: replace all `behaves same` with `behaves same as` on all the lines

File: pyalgotrading/algobulls/connection.py [hunk 16]
```diff
@@ -310,14 +409,21 @@ def backtest(self, strategy=None, start=None, end=None, instruments=None, lots=1
...
+        _error_msg_timestamps = f'\nString Format (YYYY-MM-DD HH:MM z) or an instance of type datetime.datetime for Back Testing \nString Format (HH:MM z) or an instance of type datetime.time for Real trading or Paper Trading'
         _error_msg_instruments = f'Argument "instruments" should be a valid instrument string or a list of valid instruments strings. You can use the \'get_instrument()\' method of AlgoBullsConnection class to search for instruments'
         _error_msg_mode = f'Argument "mode" should be a valid string or an enum of type StrategyMode. Possible string values can be: {get_valid_enum_names(StrategyMode)}'
 
+        initial_funds_virtual = float(initial_funds_virtual)
         if isinstance(start, str):
-            start = dt.strptime(start, '%Y-%m-%d | %H:%M')
+            try:
+                start = get_datetime_with_tz(start, trading_type.name)
+            except ValueError as ve:
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:420]: use `ex` as exception obj name and not `ve`

File: pyalgotrading/algobulls/connection.py [hunk 17]
```diff
@@ -310,14 +409,21 @@ def backtest(self, strategy=None, start=None, end=None, instruments=None, lots=1
...
+                raise ValueError(f'Error : Invalid string timestamp format for argument "start" ({ve}).\nExpected timestamp format for {trading_type.name} is [{TRADE_TYPE_DT_FORMAT_MAP[trading_type.name][KEY_DT_ZONE]}]')
         if isinstance(end, str):
-            end = dt.strptime(end, '%Y-%m-%d | %H:%M')
+            try:
+                end = get_datetime_with_tz(end, trading_type.name)
+            except ValueError as ve:
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:425]: use `ex` as exception obj name and not `ve`

File: pyalgotrading/algobulls/connection.py (hunk 14 above)
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/connection.py:386]: I think broker ID is required for RT

File: pyalgotrading/algobulls/connection.py (hunk 17 above)
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/connection.py]: done

File: pyalgotrading/algobulls/connection.py (hunk 16 above)
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/connection.py]: done

File: pyalgotrading/algobulls/connection.py (hunk 15 above)
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/connection.py]: done

File: pyalgotrading/algobulls/connection.py [hunk 18]
```diff
@@ -271,20 +279,112 @@ def get_report(self, strategy_code, trading_type, report_type, render_as_datafra
...
+                report: format and content of the report
+                html_dump: save it as a html file
+                pnl_df: dataframe containing pnl reports
+                initial_funds: initial funds to before starting the job
+            Returns:
+                Report details
+        """
+
+        order_report = None
+        # get pnl data and cleanup as per quantstats format
```
[the remark this answers is not in the document]
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/connection.py:344]: done

File: pyalgotrading/algobulls/connection.py (hunk 12 above)
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/connection.py]: done

File: pyalgotrading/algobulls/connection.py (hunk 11 above)
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/connection.py]: done

File: pyalgotrading/algobulls/api.py (hunk 10 above)
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/api.py]: done

File: pyalgotrading/algobulls/connection.py (hunk 6 above)
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/connection.py]: done

File: pyalgotrading/algobulls/connection.py (hunk 7 above)
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/connection.py]: done

File: pyalgotrading/algobulls/api.py (hunk 5 above)
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/api.py]: done

File: pyalgotrading/algobulls/connection.py [hunk 19]
```diff
@@ -310,14 +409,21 @@ def backtest(self, strategy=None, start=None, end=None, instruments=None, lots=1
...
         _error_msg_instruments = f'Argument "instruments" should be a valid instrument string or a list of valid instruments strings. You can use the \'get_instrument()\' method of AlgoBullsConnection class to search for instruments'
         _error_msg_mode = f'Argument "mode" should be a valid string or an enum of type StrategyMode. Possible string values can be: {get_valid_enum_names(StrategyMode)}'
-
+        _error_msg_broking_details = 'Argument "broking_details" should be a valid dict with valid keys. Expected keys "brokerName" or "brokerId" and "credentialParameters" '
+        initial_funds_virtual = float(initial_funds_virtual)
         if isinstance(start, str):
-            start = dt.strptime(start, '%Y-%m-%d | %H:%M')
+            try:
+                start = get_datetime_with_tz(start, trading_type.name)
+            except ValueError as ex:
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:420]: i think we might not get this error.

File: pyalgotrading/algobulls/connection.py [hunk 20]
```diff
@@ -310,14 +409,21 @@ def backtest(self, strategy=None, start=None, end=None, instruments=None, lots=1
...
+                raise ValueError(f'Error : Invalid string timestamp format for argument "start" ({ex}).\nExpected timestamp format for {trading_type.name} is [{TRADE_TYPE_DT_FORMAT_MAP[trading_type.name][KEY_DT_ZONE]}]')
         if isinstance(end, str):
-            end = dt.strptime(end, '%Y-%m-%d | %H:%M')
+            try:
+                end = get_datetime_with_tz(end, trading_type.name)
+            except ValueError as ex:
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:425]: i think we might not get this error.

**Review by pushpakdagade — APPROVED** [review]
